### DISCUSSION TITLE
Channel utilization (%)

### DISCUSSION THREAD

**Discussion by AlexZavgorodniy91** [jgromes/RadioLib#720] [General]
Hi everyone!

In our project, it is necessary to measure the channel utilization (%) of the SX1262 transceiver. 

Library provides AirTime calculation for each packet, and I guess it's possible to calculate channel utilization using this:

```
totalAirTime = airTime * packetsTransmitted;
channelUtilization = (totalAirTime / observationTime) * 100;
```

What would be the best choice to measure it? Perhaps there are other ways to do this. Can anyone suggest me the right way to do this?

Thanks!

**Comment by jgromes** [comment]
I'm not entirely sure why would you need this. First, I'm going to assume you're talking about channel utilization in time, not in frequency. If that is correct, then you can do what you suggested, for every packet you received, count its length, call `getTimeOnAir(len)` and you will know how long was taht packet on air. Add this number to a cumulative sum and when you want to know the time utilization, just divide it by the `millis()` counter, which is elapsed time since reboot.

The downside of this approach is that you will only be able to count the packets that were successfully received, so you will likely miss some and the percentage you get will be an underestimate. Also, if there is some interference in the channel (e.g. non-LoRa packets), it will not be counted to the total.

Alternatively, you could sample the RSSI at a fixed interval (e.g. couple ms), and if it is above threshold, consider the channel occupied. If not, the channel is free. The advantage is that this approach does not depend on being able to decode the packet. The downside is that all signals within your channel will be counted towards the total.

Third option I can thing of is using channel activity detection. Calling CAD in fixed intervals (e.g. couple ms) will tell you if there is an ongoing LoRa transmission or not. You can then use this information similarly as the RSSI measurement, the only difference is that CAD will not count non-LoRa signals.

Which approach is best for you depends on what exactly are you trying to achieve.

**Comment by AlexZavgorodniy91** [comment]
Thanks for the quick response!

Yes, we need to measure channel utilization in time.

Btw, how I can check CAD? Using `getIrqStatus()` or something else?

**Reply by jgromes** [reply]
See the SX126x_Channel_Activity_Detection(_Interrupt) examples.

**Reply by AlexZavgorodniy91** [reply]
I'll review this example. Thanks!

**Comment by vtjballeng** [comment]
> I'm not entirely sure why would you need this.

@jgromes, in a busy multi-channel, multi-antenna environment with many nodes and multiple gateways, this can help determine which channels are under or over utilized. 

Some of the concern here stems from the IRQ servicing, standby states, collisions/incomplete packets, and possibility that CAD doesn't catch all conditions. 

totalAirTime should account for both RX & TX. 

Elevated RSSI does offer an interesting approach but could be skewed by interference as I understand it. I'm not sure there is a perfect approach here.

At a high level, the goal is to determine the total time a given transceiver is occupied either with an RX or TX event, even an incomplete message.

**Reply by jgromes** [reply]
OK, then I guess it's really up to the definition of what should be counted. If the goal is to spread load across different channels, then I would say RSSI measurement is the best way to go, since channels with interference should be used less than channels without. I can also imagine an algorithm that would take advantage of the actual RSSI value and do something smarter than just count the time the RSSI is above/below some threshold, e.g. to prefer channels with lower noise floor.
